src/weak_memory.rs

```diff
@@ -0,0 +1,271 @@
+        global: &GlobalState,
+        is_seqcst: bool,
+        rng: &mut (impl rand::Rng + ?Sized),
+        validate: impl FnOnce() -> InterpResult<'tcx>,
+    ) -> InterpResult<'tcx, Option<ScalarMaybeUninit<Tag>>> {
+        // Having a live borrow to store_buffer while calling validate_atomic_load is fine
+        // because the race detector doesn't touch store_buffer
+        let store_buffer = self.get_store_buffer(range);
+
+        let store_elem = {
+            // The `clocks` we got here must be dropped before calling validate_atomic_load
+            // as the race detector will update it
+            let (.., clocks) = global.current_thread_state();
+            // Load from a valid entry in the store buffer
+            store_buffer.fetch_store(is_seqcst, &clocks, &mut *rng)
+        };
+
+        // Unlike in write_scalar_atomic, thread clock updates have to be done
+        // after we've picked a store element from the store buffer, as presented
+        // in ATOMIC LOAD rule of the paper. This is because fetch_store
+        // require access to ThreadClockSet.clock, which is updated by the race detector
+        validate()?;
+
+        let loaded = store_elem.map(|store_elem| {
+            let (index, clocks) = global.current_thread_state();
+            store_elem.load_impl(index, &clocks)
+        });
+        Ok(loaded)
+    }
+
+    pub fn buffered_write<'tcx>(
+        &mut self,
+        val: ScalarMaybeUninit<Tag>,
+        range: AllocRange,
+        global: &GlobalState,
+        is_seqcst: bool,
+    ) -> InterpResult<'tcx> {
+        let (index, clocks) = global.current_thread_state();
+
+        let mut store_buffer = self.get_store_buffer_mut(range);
+        store_buffer.store_impl(val, index, &clocks.clock, is_seqcst);
+        Ok(())
+    }
+}
+
+const STORE_BUFFER_LIMIT: usize = 128;
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct StoreBuffer {
+    // Stores to this location in modification order
+    buffer: VecDeque<StoreElement>,
+}
+
+impl Default for StoreBuffer {
+    fn default() -> Self {
+        let mut buffer = VecDeque::new();
+        buffer.reserve(STORE_BUFFER_LIMIT);
+        Self { buffer }
+    }
+}
+
+impl<'mir, 'tcx: 'mir> StoreBuffer {
+    /// Selects a valid store element in the buffer.
+    /// The buffer does not contain the value used to initialise the atomic object
+    /// so a fresh atomic object has an empty store buffer until an explicit store.
+    fn fetch_store<R: rand::Rng + ?Sized>(
+        &self,
+        is_seqcst: bool,
+        clocks: &ThreadClockSet,
+        rng: &mut R,
+    ) -> Option<&StoreElement> {
+        use rand::seq::IteratorRandom;
+        let mut found_sc = false;
+        // FIXME: this should be an inclusive take_while (stops after a false predicate, but
+        // includes the element that gave the false), but such function doesn't yet
+        // exist in the standard libary https://github.com/rust-lang/rust/issues/62208
```

Comment:
So -- does that mean the code below is wrong? If `take_while` doesn't have the right semantics then you just can't use `take_while`, or am I missing something?

Reply:
The code is correct. The `keep_searching` variable is there to delay the end of the iterator by one element. The boundary element still return true from the closure but sets `keep_searching` to false.

Is there a way to express this with a while loop and still keeping the iterator around without resorting to pushing things to `Vec`?

Reply:
Can you run the iterator by hand? Like, how `for` is just sugar for
```rust
let mut iter = make_iterator();
while let Some(elem) = iter.next() {
  // ...
}
```

Reply:
I could, but it consumes the iterator and makes the later `.choose(rng)` call impossible without pushing the elements to a separate Vec. I can't get a random bool as I go because that would make the probabilities geometric not uniform, plus I don't know how many candidates there are.

Reply:
I see, makes sense.

> The code is correct. The keep_searching variable is there to delay the end of the iterator by one element. The boundary element still return true from the closure but sets keep_searching to false.

Please mention that in the comment; right now it reads like it is pointing out a bug in the code ("this should be X [but isn't]").

Also, maybe the syntactic structure if this could be rearranged to make this more clear? Like, make the fn body an `else if` chain with a surrounding `keep_searching = if ...` and a single `return true` at the end.